### PR TITLE
[ONNX][Relay] Support "tf_crop_and_resize" in relay Resize op.

### PR DESCRIPTION
**Pull request by mbrookhart**

### CODE REVIEW

File: python/tvm/relay/frontend/onnx.py [hunk 1]
```diff
@@ -2675,27 +2684,55 @@ def v11_13_common(cls, inputs, size, attr, params):
...
+                coord_trans,
+                nearest_mode,
+                alpha,
+                exclude,
+                extrapolation_value,
             )
         else:
             raise NotImplementedError("Resize only supports 3, 4, or 5 dims")
-
+        print(out)
```
Comment (jwfromm) [inline python/tvm/relay/frontend/onnx.py:2735]: forgot to remove a print.

File: tests/python/frontend/onnx/test_forward.py [hunk 2]
```diff
@@ -4955,7 +4955,7 @@ def verify_eyelike(indata):
     "test_reduce_sum_keepdims_random",
     "test_reduce_sum_negative_axes_keepdims_example",
     "test_reduce_sum_negative_axes_keepdims_random",
-    "test_resize_tf_crop_and_resize",
+    # "test_resize_tf_crop_and_resize",
```
Comment (jwfromm) [inline tests/python/frontend/onnx/test_forward.py:4958]: We should delete this instead of comment right?

File: src/relay/transforms/pattern_utils.h [hunk 3]
```diff
@@ -450,6 +450,23 @@ static inline Array<Integer> ToVector(const runtime::NDArray& array) {
   return out;
 }
 
+/*!
+ * \brief Convert a NDArray with type int or float to Array<FloatImm>.
+ * \param array Input NDArray
+ * \return Converted Array.
+ */
+static inline Array<FloatImm> ToFloatVector(const runtime::NDArray& array) {
```
Comment (jwfromm) [inline src/relay/transforms/pattern_utils.h:458]: Just curious, why was this needed? I dont see how it interacts with resize.

File: python/tvm/relay/op/image/image.py [hunk 4]
```diff
@@ -125,6 +143,9 @@ def resize2d(
     size: Tuple of Int or Expr
         The out size to which the image will be resized.
 
+    roi: Tuple of Float or Expr
+        The region of interest for tf_crop_and_resize.
```
Comment (jwfromm) [inline python/tvm/relay/op/image/image.py:147]: I think itd be helpful to describe the expected shape of ROI since [2*dims] isnt obvious and is a hard requirement. Just adding it to the relay parameter comments should be enough.

**Review by jwfromm — APPROVED** [review]
A few nitpicks, overall is well structured and LGTM.

File: src/relay/transforms/pattern_utils.h (hunk 3 above)
Comment (mbrookhart) [reply src/relay/transforms/pattern_utils.h:458]: I used it in DynamicToStatic to convert the ROI argument.

File: src/relay/op/image/resize.cc [hunk 5]
```diff
@@ -215,6 +225,8 @@ bool Resize3DRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
 
   const Resize3DAttrs* param = attrs.as<Resize3DAttrs>();
   ICHECK(param != nullptr);
+  ICHECK(param->size.size() == 3);
+  ICHECK(param->roi.size() == 10);
```
Comment (masahi) [inline src/relay/op/image/resize.cc:229]: Why 10 here? Similarly, where do 6 or 8 for 1D/2D come from? Doesn't onnx doc say this number is `rank x 2`? https://github.com/onnx/onnx/blob/master/docs/Operators.md#Resize You also use only 2/4/6 of rois in topi functions for 1D/2D/3D.

Comment (mbrookhart) [reply src/relay/op/image/resize.cc:229]: Rank here includes N and C, for for 3D resize, rank is 5, and roi is len 10.

Comment (masahi) [reply src/relay/op/image/resize.cc:229]: I see. This means `roi` array is sensitive to the layout, and your change in `topi/image/resize.py` seems to support only NCW/NCHW/NCDHW. 

While the ONNX definition supports arbitrary rank and doesn't distinguish the positions of spatial axes, our op supports only certain layout + corresponding axes position. I think it is better to remove roi for batch and channel dims in the frontend (similar to how `out_size` is handled) and require `param->roi.size() == 6` here for better clarity and removing the need for dealing with roi layout in topi.

Comment (masahi) [reply src/relay/op/image/resize.cc:229]: Related to above, I wonder if it is safe hard code NCH/NCHW/NCDHW layout in the ONNX `Resize` op converter, when the spec doesn't care about layout... https://github.com/apache/tvm/blob/4c1b66f9379fed29988cefd7334f0cd4918c8e6a/python/tvm/relay/frontend/onnx.py#L2612-L2619. I remember the folklore of ONNX being tied to NCHW layout, but I don't find the official statement of that convention in the spec. 

Is this code supposed to work on models exported from TF?

Comment (mbrookhart) [reply src/relay/op/image/resize.cc:229]: > I think it is better to remove roi for batch and channel dims in the frontend (similar to how out_size is handled) and require param->roi.size() == 6 here for better clarity and removing the need for dealing with roi layout in topi.

That makes sense, I'll work on that. 

> I remember the folklore of ONNX being tied to NCHW layout, but I don't find the official statement of that convention in the spec.

Convolution definitely has to be NCHW. 

https://github.com/onnx/onnx/blob/master/docs/Operators.md#inputs-2---3

The documentation isn't as expilict about it with Resize, but I haven't found a test or an example of anything else. 

I'm looking at the code. Nearest neighbor appears to go over all axes: https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/upsample.cc#L136-L254

But bilinear is only supporting NCHW: https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/upsample.cc#L1069-L1075
